Add nix hash convert

kolloch opened the pull request:
# Motivation

This adds the `nix hash convert` command as described in https://github.com/NixOS/nix/issues/8876.

This does slightly deviate from the proposal: `--to` defaults to `sri`.

# Context

Open:

- [x] Better CLI help text with available options.
- [x] Update docs? Or which part is auto-generated which is not?
- [x] Clarify whether the base32 vs nix32 hash format name change should be part of this.
- [x] Verify that the approach to making "--from" AND "--type" AND "--to" optional depending on context is a good choice.
- [x] Verify that the "--from" parameter strategy is as desired. It is an assert rather than a format hint since it is redundant.

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

fricklerhandwerk commented on .gitignore:
```diff
@@ -144,3 +144,6 @@ result
 
 # clangd and possibly more
 .cache/
+
+# Mac OS
+.DS_Store
```
```suggestion
```

Project-level `.gitignore` is only for project-specific things. Please fix this in your user-level config.

kolloch replied:
Interesting! I didn't even know that existed...

fricklerhandwerk commented on src/nix/hash.cc:
```diff
@@ -127,10 +132,73 @@ struct CmdToBase : Command
+        });
+        addFlag({
+            .longName = "type",
+            .description = "Specify the type if it can't be auto-detected.",
```
```suggestion
            .description = "Specify the type if it's not SRI.",
```

fricklerhandwerk commented on src/nix/hash.cc:
```diff
@@ -127,10 +132,73 @@ struct CmdToBase : Command
+
+    std::string description() override
+    {
+        return "convert between different hash formats, e.g. base16 and sri.";
```
This is where we can already make it explicit, even before we have automation in place. Ideally though it would be an automatic enumeration of the supported formats that can be re-used everywhere it's needed, such as in error messages (e.g. `x was given, but must be one of %s`).

kolloch replied:
Not super sure what you mean here but I adjusted it as I though would be a good trade-off. between effort and effect.

fricklerhandwerk commented on tests/functional/hash.sh:
```diff
@@ -81,24 +81,41 @@ rm $TEST_ROOT/hash-path/hello
-# Conversion.
+# Conversion with `nix hash` `nix-hash` and `nix hash convert`
 try3() {
+    # $1 = hash type
```
Would be good to also test the SRI in/out cases.

fricklerhandwerk reviewed (commented):
Cool, thanks a lot for helping out! Would be great if you could add a release note. And ideally a deprecation warning on the existing command, but this can be done in a follow-up not to stall progress here if you're short on time.

tomberek commented:
@kolloch , for your questions and from Nix team

-  [x] Update docs? Or which part is auto-generated which is not?
    - yes, and please add docs
- [x] Clarify whether the base32 vs nix32 hash format name change should be part of this.
    - yes
- [x]  Verify that the approach to making "--from" AND "--type" AND "--to" optional depending on context is a good choice.
    - yes, if it can be unambiguous
    - please consider changing "--type" into "--algo" (thoughts?), we updated the original issue
 - [x] Verify that the "--from" parameter strategy is as desired. It is an assert rather than a format hint since it is redundant.
    - yes
- [ ] more tests are always wonderful to clarify expected behavior (eg: what was in the comment in the parent issue)

kolloch commented:
I *think* I addressed all the comments.

In addition, I did rename `HashType` to `HashAlgorithm` so that the type matches the arguement names of the CLI and the `builtin`.

I did NOT yet remove the need to specify the `nix-command` experiment feature.

It would be nice to get this merged in the current form plus eventually needed bug fixes, of course. Meaning: it might still be worth it to add some more docs somewhere but I'd like to do future works in separate merge requests so this doesn't grow unboundedly.

Thank you!

kolloch commented:
> How exactly does this deviate from the proposal though?

The `--to` format parameter is optional and defaults to SRI.

kolloch commented:
Ah, I noticed that my IDE reindented some lines close to some renames. I hope that's fine. It would be quite some pain to undo.

kolloch commented:
And some further fixes in code that wasn’t executed in `make installcheck`

kolloch commented:
@tomberek anything else?

kolloch commented on src/libexpr/tests/libnixexpr-tests:
maybe we should add this to .gitignore?

kolloch commented on doc/manual/src/release-notes/rl-next.md:
<img width="860" alt="image" src="https://github.com/NixOS/nix/assets/339354/0ae3612a-a172-4ca1-a455-b82e08a47ede">

kolloch commented:
Thank you for checking and finding the added files, @tomberek!

I interactively rebased to make sure that the binary doesn't pollute the history.

Not sure what this is about and if I can do something about it:

https://github.com/NixOS/nix/actions/runs/7112656418/job/19363023090?pr=9452
<img width="914" alt="image" src="https://github.com/NixOS/nix/assets/339354/e355cb93-0ccf-4e7e-884f-e786987d890a">

kolloch commented:
Somehow it still shows I need to address some review comments but I don't see which...

Ericson2314 commented:
Sorry I dropped the ball on this. @tomberek thank you for picking up after me.

kolloch commented:
Hurray! Thanks @tomberek
